File: content/snippets/scala3/design_patterns_interpreter.scala
// Interpreter
trait Expression:
    def interpret(ctx: String): Boolean

class TerminalExpression(literal: String) extends Expression:
    def interpret(ctx: String): Boolean = ctx.contains(literal)

class OrExpression(expr1: Expression, expr2: Expression) extends Expression:
    def interpret(ctx: String): Boolean = expr1.interpret(ctx) || expr2.interpret(ctx)

class AndExpression(expr1: Expression, expr2: Expression) extends Expression:
    def interpret(ctx: String): Boolean = expr1.interpret(ctx) && expr2.interpret(ctx)

@main def main(): Unit =
    val c1: Expression = new TerminalExpression("cat")
    val c2: Expression = new TerminalExpression("dog")
    val expr: Expression = new OrExpression(c1, c2)
    
    println(expr.interpret("cat"))
    println(expr.interpret("bird"))

File: content/snippets/python/design_patterns_interpreter.py
from abc import ABC, abstractmethod

class Expression(ABC):
    @abstractmethod
    def interpret(self, ctx: str) -> bool:
        pass

class TerminalExpression(Expression):
    def __init__(self, literal: str):
        self.literal = literal
    
    def interpret(self, ctx: str) -> bool:
        return self.literal in ctx

class OrExpression(Expression):
    def __init__(self, expr1: Expression, expr2: Expression):
        self.expr1 = expr1
        self.expr2 = expr2
    
    def interpret(self, ctx: str) -> bool:
        return self.expr1.interpret(ctx) or self.expr2.interpret(ctx)

class AndExpression(Expression):
    def __init__(self, expr1: Expression, expr2: Expression):
        self.expr1 = expr1
        self.expr2 = expr2
    
    def interpret(self, ctx: str) -> bool:
        return self.expr1.interpret(ctx) and self.expr2.interpret(ctx)

c1 = TerminalExpression("cat")
c2 = TerminalExpression("dog")
expr = OrExpression(c1, c2)

print(expr.interpret("cat"))
print(expr.interpret("bird"))
